tsl/src/nodes/vector_agg/function/functions.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+
+#include <limits.h>
+
+#include <postgres.h>
+
+#include <common/int.h>
+#include <utils/date.h>
+#include <utils/float.h>
+#include <utils/fmgroids.h>
+#include <utils/fmgrprotos.h>
+
+#include "functions.h"
+
+#include "compat/compat.h"
+
+/*
+ * Aggregate function count(*).
+ */
+typedef struct
+{
+	int64 count;
+} CountState;
+
+static void
+count_init(void *agg_state)
+{
+	CountState *state = (CountState *) agg_state;
+	state->count = 0;
+}
+
+static void
+count_emit(void *agg_state, Datum *out_result, bool *out_isnull)
+{
+	CountState *state = (CountState *) agg_state;
+	*out_result = Int64GetDatum(state->count);
+	*out_isnull = false;
+}
+
+static void
+count_star_const(void *agg_state, Datum constvalue, bool constisnull, int n,
+				 MemoryContext agg_extra_mctx)
+{
+	CountState *state = (CountState *) agg_state;
+	state->count += n;
+}
+
+VectorAggFunctions count_star_agg = {
+	.state_bytes = sizeof(CountState),
+	.agg_init = count_init,
+	.agg_const = count_star_const,
+	.agg_emit = count_emit,
+};
+
+/*
+ * Aggregate function count(x).
+ */
+static void
+count_any_const(void *agg_state, Datum constvalue, bool constisnull, int n,
+				MemoryContext agg_extra_mctx)
+{
+	if (constisnull)
+	{
+		return;
+	}
+
+	CountState *state = (CountState *) agg_state;
+	state->count += n;
+}
+
+static void
+count_any_vector(void *agg_state, const ArrowArray *vector, const uint64 *filter,
+				 MemoryContext agg_extra_mctx)
+{
+	CountState *state = (CountState *) agg_state;
+	const int n = vector->length;
+	const uint64 *restrict validity = (uint64 *) vector->buffers[0];
+	/* First, process the full words. */
+	for (int i = 0; i < n / 64; i++)
```

Comment:
They still have the uint64 support, so it should work just as well. Probably slower, I didn't test.

Reply:
Yes, the compiler has uint64 support, but it is not a native word-length type on those machines so they need to emulate 64-bit integers.

We do build for 32 bit, but it is probably not a platform we optimize for in terms of performance. It might be good to point out that we currently don't care about optimizing for 32 bit specifically.